overlord/servicestate/quota_handlers.go

```diff
@@ -37,6 +40,302 @@ import (
+	default:
+		err = fmt.Errorf("unknown action %q requested", qc.Action)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	t.SetStatus(state.DoneStatus)
+
+	return nil
+}
+
+func doCreateQuota(args quotaControlArgs) error {
+	st := args.St
+	action := args.Action
+	allGrps := args.AllGrps
+
+	// make sure the group does not exist yet
+	if _, ok := allGrps[action.QuotaName]; ok {
+		return fmt.Errorf("group %q already exists", action.QuotaName)
+	}
+
+	// make sure the specified snaps exist and aren't currently in another group
+	if err := validateSnapForAddingToGroup(st, action.AddSnaps, action.QuotaName, allGrps); err != nil {
+		return err
+	}
+
+	// make sure the memory limit is not zero
+	if action.MemoryLimit == 0 {
+		return fmt.Errorf("internal error, MemoryLimit option is mandatory for create action")
+	}
+
+	grp, allGrps, err := createQuotaImpl(st, action, allGrps)
+	if err != nil {
+		return err
+	}
+
+	// ensure the snap services with the group
+	opts := &ensureSnapServicesForGroupOptions{
+		allGrps: allGrps,
+	}
+	return ensureSnapServicesForGroup(st, grp, opts, args.Meter, args.PerfTimings)
+}
+
+func createQuotaImpl(st *state.State, action *QuotaControlAction, allGrps map[string]*quota.Group) (*quota.Group, map[string]*quota.Group, error) {
+	// make sure that the parent group exists if we are creating a sub-group
+	var grp *quota.Group
+	var err error
+	updatedGrps := []*quota.Group{}
+	if action.ParentName != "" {
+		parentGrp, ok := allGrps[action.ParentName]
+		if !ok {
+			return nil, nil, fmt.Errorf("cannot create group under non-existent parent group %q", action.ParentName)
+		}
+
+		grp, err = parentGrp.NewSubGroup(action.QuotaName, action.MemoryLimit)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		updatedGrps = append(updatedGrps, parentGrp)
+	} else {
+		// make a new group
+		grp, err = quota.NewGroup(action.QuotaName, action.MemoryLimit)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	updatedGrps = append(updatedGrps, grp)
+
+	// put the snaps in the group
+	grp.Snaps = action.AddSnaps
+	// update the modified groups in state
+	newAllGrps, err := patchQuotas(st, updatedGrps...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return grp, newAllGrps, nil
+}
+
+type quotaControlArgs struct {
+	St          *state.State
+	AllGrps     map[string]*quota.Group
+	Action      *QuotaControlAction
+	Meter       progress.Meter
+	PerfTimings *timings.Timings
+}
+
+func doRemoveQuota(args quotaControlArgs) error {
+	st := args.St
+	action := args.Action
+	allGrps := args.AllGrps
+
+	// make sure the group exists
+	grp, ok := allGrps[action.QuotaName]
+	if !ok {
+		return fmt.Errorf("group %q does not exist", action.QuotaName)
+	}
+
+	// make sure some of the options are not set, it's an internal error if
+	// anything other than the name and action are set for a removal
+	if action.ParentName != "" {
+		return fmt.Errorf("internal error, ParentName option does not make sense with remove action")
```

Comment:
`s/does not make sense/cannot be used/`? here and the next 2 errors that follow

Reply:
done